### ISSUE TITLE
no server connection

### ISSUE THREAD

**Issue by lars20070** [Bioconductor/UniProt.ws#9]
A simple script
```
library(UniProt.ws)
up <- UniProt.ws(taxId=9606)
```
results in the error
```
Error in file(con, "r") : cannot open the connection
In addition: Warning message:
In file(con, "r") : 
InternetOpenUrl failed: ´A connection with the server could not be established´
```
I can install any R package i.e. internet on my machine is fine. Any suggestions how to continue to debug?

**Comment by lshep** [comment]
Could you please try again and if you continue to have issues please provide your `sessionInfo()` . It could have been a temporary issue with the UniProt server not necessarily your internet connection.   If the server was temporarily down, it would be unreachable.  I just tried your example using both Bioc release 3.8 and Bioc devel 3.9  and did not run into any ERRORs.

**Comment by lars20070** [comment]
@lshep Thanks for the quick reply. The UniProt server seems no problem. And the internet connection neither.

![error](https://user-images.githubusercontent.com/6149805/55238983-09583c80-5236-11e9-8386-47629b0011fc.png)

**Comment by lshep** [comment]
Do you have the option of updating your version of R and to the latest version of the package?  This would be recommended.

**Comment by lars20070** [comment]
I upgraded to R 3.5.3, but cannot install `UniProt.ws`. Guess this is related.

In the browser I can access
`https://cran.rstudio.com/src/contrib/PACKAGES`
And I can ping from the command prompt.

![Screenshot from 2019-03-29 15-39-58](https://user-images.githubusercontent.com/6149805/55240340-014dcc00-5239-11e9-84c9-767952c52797.png)

And I can ping in R.

![ping](https://user-images.githubusercontent.com/6149805/55241879-5b03c580-523c-11e9-9a0e-298120d73fbc.png)

**Comment by lars20070** [comment]
Could this be `https`related?

**Comment by lshep** [comment]
Please use BiocManager to install packages.   
```
install.package("BiocManager") 
install("UniProt.ws")
library(UniProt.ws)
```

**Comment by lars20070** [comment]
![install](https://user-images.githubusercontent.com/6149805/55242258-25131100-523d-11e9-88e2-805009d81298.png)

**Comment by lshep** [comment]
Sorry 
```
library(BiocManager)
install("UniProt.ws")
library(UniProt.ws)
```

**Comment by lars20070** [comment]
Installation worked. UniProt access not.

![sessiojn](https://user-images.githubusercontent.com/6149805/55242513-ac608480-523d-11e9-9bbd-b4b72b6bbb90.png)

**Comment by lars20070** [comment]
I can ping `uniprot.org`.

![uni](https://user-images.githubusercontent.com/6149805/55243915-6bb63a80-5240-11e9-9f7e-b756e6a78f4c.png)

**Comment by lshep** [comment]
We will have to investigate more.

**Comment by lars20070** [comment]
Would be good. Thanks.

We have been using `UniProt.ws` successfully for months. And I do not want to switch to a local version.

**Comment by mtmorgan** [comment]
The specific query is

```
url = "https://www.uniprot.org/uniprot/?query=organism:9606&format=tab&columns=id"
```
Can you reach that from your browser? Using `readLines(url)` ? `readLines()` uses `file()` to read from a connection, and `?file` mentions `getOption("url.method")` What value do you have for
```
getOption("url.method")
```
There are some alternatives described on `?download.file` and it might help to try these... the url is `https:`, which in particular might cause problems...

**Comment by lars20070** [comment]
I can access `url` from the browser but *not* with `readLines()`.
`getOption("url.method")` is `NULL`. The [download.file](https://stat.ethz.ch/R-manual/R-devel/library/utils/html/download.file.html) page does indeed list the solution.
```
options(url.method="libcurl")
```
seems to unblock `https` access.

**Comment by lars20070** [comment]
@mtmorgan @lshep Thank you for the good support.

Could you test and set this option automatically if necessary?
Or can you set this `url.method` by default? It seems to work in Windows, Ubuntu and macOS.

**Comment by lars20070** [comment]
Despite the new `url.method` option, I still run into server connection problems under Windows. The errors are sporadic, not very reproducible. Only every now and then. 

![chunk](https://user-images.githubusercontent.com/6149805/55339912-e5019780-54a3-11e9-8df7-41d26ad7b520.png)

Under macOS and Ubuntu, my scripts are stable.

**Comment by mtmorgan** [comment]
The url.method option is not something that a package should set; the precise setting depends on the user's particular computer.

Did `options(url.method="libcurl")` work for you initially, and then stop working, or did it work with `readLines(url)` but not with the actual function call?

**Comment by Kayla-Morrell** [comment]
@lars20070 - Just checking in to see if this issue has been resolved or not.

**Comment by lars20070** [comment]
@Kayla-Morrell Thanks for keeping track of this issue. I no longer have access to a Windows machine and cannot help you with further testing. Sorry.

**Comment by vw087** [comment]
Hello @Kayla-Morrell   I use Window and I keep running in to same issue. I was wondering whether you were able to find a solution to overcome this?
> sessionInfo()
R version 3.5.3 (2019-03-11)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] UniProt.ws_2.29.0   BiocGenerics_0.28.0 RCurl_1.98-1.1      RSQLite_2.1.2      

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.4           pillar_1.4.2         compiler_3.5.3       BiocInstaller_1.30.0
 [5] BiocManager_1.30.10  dbplyr_1.4.2         prettyunits_1.0.2    remotes_2.1.0       
 [9] bitops_1.0-6         tools_3.5.3          pkgbuild_1.0.6       zeallot_0.1.0       
[13] digest_0.6.22        bit_1.1-14           memoise_1.1.0        BiocFileCache_1.6.0 
[17] tibble_2.1.3         pkgconfig_2.0.3      rlang_0.4.1          cli_1.1.0           
[21] DBI_1.0.0            rstudioapi_0.10      curl_4.3             stringr_1.4.0       
[25] withr_2.1.2          dplyr_0.8.3          httr_1.4.1           S4Vectors_0.20.1    
[29] vctrs_0.2.0          IRanges_2.16.0       rappdirs_0.3.1       rprojroot_1.3-2     
[33] stats4_3.5.3         bit64_0.9-7          tidyselect_0.2.5     glue_1.3.1          
[37] Biobase_2.42.0       R6_2.4.1             processx_3.4.1       AnnotationDbi_1.44.0
[41] callr_3.3.2          purrr_0.3.2          blob_1.2.0           magrittr_1.5        
[45] ps_1.3.0             backports_1.1.5      assertthat_0.2.1     stringi_1.4.3       
[49] crayon_1.3.4        
>

**Comment by TomSmithCGAT** [comment]
TL;DR - @vw087 and others with the same issue. Try re-installing `openssl` R package: `install.packages('openssl')`

I was having the same issue as described above 

```
library(UniProt.ws)
up <- UniProt.ws(taxId=9606)
```
`URL 'https://www.uniprot.org/uniprot/?query=organism:9606&format=tab&columns=id': status was 'SSL connect error'Error in file(con, "r") : 
  cannot open the connection to 'https://www.uniprot.org/uniprot/?query=organism:9606&format=tab&columns=id'`

And was unable to solve it by setting e.g `options(url.method="libcurl")` as mentioned about. 

Pointing `readLines` to the same url, the error mentioned `status was 'SSL connect error'Error in file(con, "r")`. With `options(download.file.method="wget")`, the error also stated: `sslv3 alert handshake failure`. 

A quick google suggested the issue was probably related to OpenSSL. Updating `openssl` via `apt install` didn't help. Re-installing the `openssl` R package (`install.packages('openssl')`) solved the issue. If I knew more about any of this, I might be able to venture a guess at why? :man_shrugging:

**Comment by lshep** [comment]
What OS are you using? Could you share your `sessionInfo()`

**Comment by TomSmithCGAT** [comment]
Just to make life fun, I'm now getting the same error messages again so the above clearly didn't solve it permanently. 

Via command line curl , I get the following
```
$ curl --verbose https://www.uniprot.org/uniprot/?query=organism:9606&format=tab&columns=id
Trying 193.62.193.81:443...
* TCP_NODELAY set
* Connected to www.uniprot.org (193.62.193.81) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS alert, handshake failure (552):
* error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure
* Closing connection 0
curl: (35) error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure
```

This can be resolved adding the option `--ciphers DEFAULT@SECLEVEL=1`, as suggested [here](https://unix.stackexchange.com/questions/590776/curl-gives-ssl-routinesssl3-read-bytessslv3-alert-handshake-failure). 

This looks to be a specific issue with the default security level for curl with Ubuntu 20.04, a bug in OpenSSL 1.1.1 and the configuration of the Uniprot server - a deeper dive [here - Ensembl #427](https://github.com/Ensembl/ensembl-rest/issues/427#issue-614497457)

From [here](https://msmith.de/2020/10/02/httr-curl-ubuntu-20-04.html), I've established that it's possible to set global httr configuration options `set_config(config(ssl_cipher_list = "DEFAULT@SECLEVEL=1"))` such that `httr::GET` works fine with the uniprot url. However, I can't see how to similarly set global options for e.g `curl::curl_fetch_memory` so this doesn't solve my issue when calling `up <- UniProt.ws(taxId=9606)`.

For now, I'll try [re-configuring the default SSL security level](https://askubuntu.com/questions/1233186/ubuntu-20-04-how-to-set-lower-ssl-security-level), but it would be really nice to know a way around this within R instead.

I note my issue is also slightly different from the original issue. Would you like me to open up a new issue rather re-opening this one?

My OS and session info:

Ubuntu 20.04 LTS

```
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8    LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] UniProt.ws_2.28.0   BiocGenerics_0.34.0 RCurl_1.98-1.2      RSQLite_2.2.0      

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.5           pillar_1.4.6         compiler_4.0.3       dbplyr_1.4.4         R.methodsS3_1.8.1   
 [6] bitops_1.0-6         R.utils_2.10.1       tools_4.0.3          digest_0.6.27        bit_4.0.4           
[11] memoise_1.1.0        BiocFileCache_1.12.1 lifecycle_0.2.0      tibble_3.0.3         pkgconfig_2.0.3     
[16] rlang_0.4.8          DBI_1.1.0            rstudioapi_0.11      curl_4.3             dplyr_1.0.2         
[21] httr_1.4.2           rappdirs_0.3.1       generics_0.0.2       vctrs_0.3.4          S4Vectors_0.26.1    
[26] IRanges_2.22.2       stats4_4.0.3         bit64_4.0.5          tidyselect_1.1.0     glue_1.4.2          
[31] Biobase_2.48.0       R6_2.4.1             AnnotationDbi_1.50.3 purrr_0.3.4          blob_1.2.1          
[36] magrittr_1.5         ellipsis_0.3.1       assertthat_0.2.1     crayon_1.3.4         R.oo_1.24.0
```

**Comment by TomSmithCGAT** [comment]
Just to confirm, changing the opensll config file `/usr/lib/ssl/openssl.cnf` as described [here](https://askubuntu.com/a/1279197) works, with the obvious downside that this entails lowering the SSL security level to 1 for all applications using `openssl`, despite the increased security level being an intentional change for Ubuntu 20.04 (see [https://bugs.launchpad.net](https://bugs.launchpad.net/ubuntu/+source/openssl/+bug/1864689)).

For anyone else getting the error below with Ubuntu >= 20.04, this is the only solution I have found.

```URL 'https://www.uniprot.org/uniprot/?query=organism:9606&format=tab&columns=id': status was 'SSL connect error' Error in file(con, "r") : cannot open the connection to 'https://www.uniprot.org/uniprot/?query=organism:9606&format=tab&columns=id'```
